module-system/sov-modules-core/src/storage/scratchpad.rs

```diff
@@ -460,6 +479,69 @@ impl<'a, C: Context> StateReaderAndWriter for AccessoryWorkingSet<'a, C> {
     }
 }
 
+/// Module for archival state
+pub mod archival_state {
+    use super::*;
+
+    /// Enum for tagging the specific type of storage.
+    enum ArchivalStore<S: Storage> {
+        /// JMT store
+        Jmt(S),
+        /// Accessory store
+        Accessory(S),
+    }
+
+    /// Archival working set
+    pub struct ArchivalWorkingSet<C: Context> {
+        delta: ArchivalStore<<C as Spec>::Storage>,
+        archival_version: Version,
+        witness: <<C as Spec>::Storage as Storage>::Witness,
+    }
+
+    impl<C: Context> ArchivalWorkingSet<C> {
+        /// Creates a new [`ArchivalWorkingSet`] instance backed by the given [`Storage`]. and [`Version`]
+        /// For jmt state access
+        pub fn new(inner: &<C as Spec>::Storage, version: Version) -> Self {
+            Self {
+                delta: ArchivalStore::Jmt(inner.clone()),
+                archival_version: version,
+                witness: Default::default(),
+            }
+        }
+
+        /// Creates a new [`ArchivalWorkingSet`] instance backed by the given [`Storage`]. and [`Version`]
+        /// For accestory state access
+        pub fn new_accessory(inner: &<C as Spec>::Storage, version: Version) -> Self {
+            Self {
+                delta: ArchivalStore::Accessory(inner.clone()),
+                archival_version: version,
+                witness: Default::default(),
+            }
+        }
+    }
+
+    impl<C: Context> StateReaderAndWriter for ArchivalWorkingSet<C> {
+        fn get(&mut self, key: &StorageKey) -> Option<StorageValue> {
+            match &self.delta {
+                ArchivalStore::Jmt(jmt_store) => {
+                    jmt_store.get(key, Some(self.archival_version), &self.witness)
+                }
+                ArchivalStore::Accessory(accessory_store) => {
+                    if !cfg!(feature = "native") {
+                        None
+                    } else {
+                        accessory_store.get_accessory(key, Some(self.archival_version))
+                    }
+                }
+            }
+        }
+
+        fn set(&mut self, _key: &StorageKey, _value: StorageValue) {}
```

Comment:
I think `set` also should be supported. So the user can run arbitrary logic against the archival state.

Reply:
This will work only per http request and won't be committed. The idea is that if you have a query method that writes to a state and then reads the same state the changes should be visible inside this single query and discarded after.

Reply:
yup, set can be supported as well - one of the reasons for making it at workingset level was to let the archival state be used to "simulate" transactions by using set, but for the immediate purpose, wanted to just support it at the RPC level

Reply:
> I think set also should be supported. So the user can run arbitrary logic against the archival state.

I agree, that it is nice to have feature, but in reality, how many queries there are, which actually write to state? 
I just want to prevent us to adding support for something, that won't be actually used, but we will keep burden of maintaining it. Sometimes limiting functionality can be benefitial.

I also see, how simulating feature definitely will be needed, but I am not sure, that it makes sense to simulate something on archival state.

Reply:
will add the set functionality in another PR for simulating module calls on the WorkingSet (without committing).

Reply:
Changing state inside a query is a standard usecase.